Feature: assemble matrix from vmult utilizing sparsity pattern.

jodlbauer opened the pull request:
As this topic came up recently on the mailing list, I decided to refurbish my existing implementation a bit and start utilizing the sparse structure as suggested by @bangerth.

It is still work in progress but does its job already. I will add some examples and documentation further on.

Whats still missing is the distributed case.

jodlbauer commented:
In my current application, the number of required vmults is around 60 for a 2d-Q1 vector valued system, and 30 for scalar valued. The numbers are almost independent of the refinement.
Tested on a globally refined grid, I expect higher number in case of adaptively refined meshes.

The resulting matrices agree perfectly with the assembled ones. I will add some tests that verify that.

peterrum commented:
Dear @jodlbauer! Thanks for you contribution. I have a couple of general questions.

Your goal (and the goal of the [mailing list discussion](https://groups.google.com/forum/#!topic/dealii/Di_cTexTxXo) - if this is the one you meant) is to compute the compute the system matrix based on a linear matrix-free operator, am I right? 

In theory you can do it simply by applying your matrix-free operator to standard basis vector. Which might become quite expensive since the needed number of global application of the operator would equal to the number of DoFs. To improve this situation, you are suggesting not to use standard basis vector but add - smartly - multiple `1`s. By graph coloring you are ensuring that there is not a mix-up. I think this approach is quite interesting! I like it. An extension to parallel would be indeed good!

We (@kronbichler and I) are currently working on a new namespace `MatrixFreeUtils` (see also the discussion in https://github.com/dealii/dealii/pull/9343#issuecomment-600338945), where we would like to collect utility functions useful in the context of `MatrixFree`. The first function I will contribute is the consistent computation of the diagonal of a matrix (given a matrix-free cell integral - and extension to DG with face and boundary integral is planed). The next function I plan to contribute is the computation of a matrix again based on cell integral (extension to DG is planed).

If you would be interested, you could add your function to that new namespace. Your function computes as my function the matrix but in a completely different way. In my case the user has to provide cell integral; in your implementation the user can provide the complete operator. Latter is not very intrusive and might be applied to an existing - complicated - matrix-free operator practically out of the box. That's why I would love to see it as an optional implementation in that namespace.

What do you think? I will try out you implementation in the next days.

jodlbauer commented:
The main goal for me was mainly a debugging and testing tool for matrix-free implementations. That way i could see more clearly what an operator did (wrong) or replace a matrix-free solver/smoother by more expensive matrix-based variants for testing.

I have moved the implementation now to the matrix-free folder and added some documentation and tests.

peterrum commented on include/deal.II/matrix_free/assemble_operator.h:
```diff
@@ -0,0 +1,278 @@
+     * @param <tt>dummy_domain</tt> a vector used to initialize vectors in the domain of <tt>op</tt> by copying the structure of
+     <tt>dummy_domain</tt>.
+     * The type <tt>DomainVector</tt> needs to provide reinit(const
+     DomainVector&, bool=false) to create a zero vector.
```
I don't like this dummy approach. Cannot we require that the operator provides also `initialize_dof_vector(vec)`?

peterrum replied:
I guess if you want to support non-symmetrical matrices, you could require `initialize_dof_vector(vec, dof_handler_index)` and then you could provide two indices one for the source vector and one for the destination vector.

jodlbauer replied:
The problem with the `initialize_dof_vector` approach is, that each Operator type (MatrixFree, LinearOperator, ...) would need to have such a method.

Using the dummy vectors, each used vector class only needs to be able to copy its structure to another instance, which I think is more flexible. That way, it is the users responsibility to provide vectors with correct block/parallel partitioning, such that they fit the Operator class.

peterrum commented on include/deal.II/matrix_free/assemble_operator.h:
```diff
@@ -0,0 +1,278 @@
+          // Apply to operator. image = sum A*e_i for i in entries.
+          RangeVector image;
+          image.reinit(dummy_range, false);
+          op.vmult(image, e);
```
You are assuming *non-symmetrical* matrices?

jodlbauer replied:
Operator might be non-symmetric (A != A^T but also n_rows != n_cols).
This is necessary because BlockSparsityPatterns do not provide begin(row)/end(row) methods, so we have to handle each block separately.

peterrum commented on include/deal.II/matrix_free/assemble_operator.h:
```diff
@@ -0,0 +1,278 @@
+          entries.reserve(pattern.max_entries_per_row());
+          for (vertices_size_type i = 0; i < pattern.n_cols(); i++)
+            if (cache->color_vec[i] == color)
+              entries.push_back(i);
```
This info is already in entries_per_color, isn't it?

jodlbauer replied:
Indeed, I messed up a bit in the caching approach. Will fix that.

peterrum commented on include/deal.II/matrix_free/assemble_operator.h:
```diff
@@ -0,0 +1,278 @@
+          // column in entry can store a value.
+          for (vertices_size_type r = 0; r < pattern.n_rows(); r++)
+            {
+              if (image[r] != 0.0)
```
Cannot you pre-compute the indices based on the color?

jodlbauer replied:
I was not sure if I should add that, as it requires to store n_colors * n_rows integers.

peterrum commented on include/deal.II/matrix_free/assemble_operator.h:
```diff
@@ -0,0 +1,278 @@
+                    auto c2 = it2->column();
+                    if (c1 > c2)
+                      std::swap(c1, c2);
+                    edges.insert(std::make_pair(c1, c2));
```
I think this line limits you to symmetric-sized matrices. I would not swap and rather make the index sets disjunct:
```suggestion
                    edges.insert(std::make_pair(c1, c2+pattern.n_rows())); // or similar
```

jodlbauer replied:
Both c1 and c2 refer to column indices, so swapping is fine. It's only to avoid duplicate entries in the undirected graph.

peterrum commented on tests/matrix_free/assemble_operator_01.cc:
```diff
@@ -0,0 +1,82 @@
+  FullMatrix<double> M;
+  M.reinit(6, 6);
+  for (auto p : pattern)
+    M(p.row(), p.column()) = std::sin(123.0 + double(p.row() * p.column()));
```
We do not use C-style casts.

jodlbauer replied:
Fixed.

peterrum commented:
@jwitte08 You were asking about graph coloring a long time ago. The code might be interesting for you.

jodlbauer commented on include/deal.II/matrix_free/assemble_operator.h:
```diff
@@ -0,0 +1,278 @@
+      std::vector<vertices_size_type> color_vec;
+      boost::iterator_property_map<vertices_size_type *, vertex_index_map>
+                                                   coloring;
+      vertices_size_type                           num_colors;
```
color_vec maps each column index to a color, so color_vec.size() == n_cols.
I guess we could compute num_colors by max(color_vec)+1.

peterrum replied:
OK!

jodlbauer commented on include/deal.II/matrix_free/assemble_operator.h:
```diff
@@ -0,0 +1,278 @@
+              typename Operator,
+              typename Pattern,
+              typename DomainVector,
+              typename RangeVector>
```
I guess so. Matrix could be any sparse or full-matrix type, Operator anything that has a vmult method. Domain/RangeVector are necessary to know which types Operator.vmult takes.
Only Pattern might be obtained from Matrix. However, this causes problems with Trilinos SparseMatrix, FullMatrix, etc ...

peterrum replied:
OK. Thanks for the explanation!

I would then replace only `DomainVector` and `RangeVector` by `VectorType`. Or are there any cases where these two do not match?

jodlbauer replied:
It is rare, but appeared sometimes in my implementations that an operator takes a blockvector and returns a non-block vector. The templates are deduced automatically, so I did not care too much about distinguishing the two types.
Anyway, if desired, I can use VectorType for both.

jodlbauer commented on include/deal.II/matrix_free/assemble_operator.h:
```diff
@@ -0,0 +1,278 @@
+      Graph                           graph;
+      std::vector<vertices_size_type> color_vec;
+      boost::iterator_property_map<vertices_size_type *, vertex_index_map>
+                                                   coloring;
```
Done.

jodlbauer replied:
Done.

jodlbauer commented on include/deal.II/matrix_free/assemble_operator.h:
```diff
@@ -0,0 +1,278 @@
+     pattern.
+     *
+     * @param <tt>output</tt> resulting matrix. Needs to provide a function set(row, column, value) that allows to write to each entry of the
+     sparsity pattern.
```
Fixed the <tt> things and capitalization. Regarding the indentation script: I use a different width setting in my personal formatter, and this seems to mess up the comments and docs. I will fix those at the very end.

jodlbauer commented on include/deal.II/matrix_free/assemble_operator.h:
```diff
@@ -0,0 +1,278 @@
+     * The type <tt>DomainVector</tt> needs to provide reinit(const
+     DomainVector&, bool=false) to create a zero vector.
+
+     * @param <tt>dummy_range</tt> see <tt>dummy_domain</tt>.
```
I will add a bit more info.

jodlbauer commented on include/deal.II/matrix_free/assemble_operator.h:
```diff
@@ -0,0 +1,278 @@
+                    AssertDimension(r, it->row());
+                    AssertDimension(r, it2->row());
+                    auto c1 = it->column();
+                    auto c2 = it2->column();
```
I'd like to have c1 < c2 later using swap, so const is not possible.

peterrum replied:
Isn't this what you want:
```suggestion
                    edges.insert({std::min(c1, c2),std::max(c1, c2)});
```

jodlbauer replied:
Ok.

jodlbauer commented on include/deal.II/matrix_free/assemble_operator.h:
```diff
@@ -0,0 +1,278 @@
+    assemble_operator(
+      Matrix &                     output,
+      const Operator &             op,
+      const Pattern &              pattern,
```
Originally, I used FullMatrix, as it was only a debug tool for small matrices.
Querying the pattern from the matrix would require different methods for SparseMatrix and Trilinos::SparseMatrix. Thats why I opted for an additional parameter.

jodlbauer commented on include/deal.II/matrix_free/assemble_operator.h:
```diff
@@ -0,0 +1,278 @@
+    void
+    assemble_operator(
+      Matrix &                     output,
+      const Operator &             op,
```
Sure. Which names would you suggest? I'd like to rename op --> operator, but C++ is not happy about this ...

peterrum replied:
Maybe: `linear_operator`?

jodlbauer replied:
Done.

peterrum commented on include/deal.II/matrix_free/assemble_operator.h:
```diff
@@ -0,0 +1,278 @@
+      boost::iterator_property_map<vertices_size_type *, vertex_index_map>
+                                                   coloring;
+      vertices_size_type                           num_colors;
+      std::vector<std::vector<vertices_size_type>> entries_per_color;
```
What about this?

jodlbauer replied:
Missed that. Whats the benefit of using that storage scheme? It feels like a lot more space for errors.
If we precompute the indices (as indicated in another comment), this would become obsolete.

peterrum commented:
I have written a small program using your function:

```cpp
#include <deal.II/base/convergence_table.h>

#include <deal.II/dofs/dof_handler.h>
#include <deal.II/dofs/dof_tools.h>

#include <deal.II/fe/fe_q.h>
#include <deal.II/fe/fe_system.h>
#include <deal.II/fe/mapping_q.h>

#include <deal.II/grid/grid_generator.h>
#include <deal.II/grid/tria.h>

#include <deal.II/lac/dynamic_sparsity_pattern.h>
#include <deal.II/lac/full_matrix.h>
#include <deal.II/lac/sparse_matrix.h>
#include <deal.II/lac/sparsity_pattern.h>
#include <deal.II/lac/vector.h>

#include <deal.II/matrix_free/assemble_operator.h>
#include <deal.II/matrix_free/fe_evaluation.h>
#include <deal.II/matrix_free/matrix_free.h>

#include <deal.II/numerics/vector_tools.h>

using namespace dealii;

template <int dim,
          int fe_degree,
          int n_points,
          int n_components,
          typename Number>
class LinearOpertor
{
public:
  LinearOpertor(const MatrixFree<dim, Number> &matrix_free)
    : matrix_free(matrix_free)
  {}

  template <typename VectorType>
  void
  vmult(VectorType &dst, const VectorType &src) const
  {
    FEEvaluation<dim, fe_degree, n_points, n_components, Number> phi(
      matrix_free);

    matrix_free.template cell_loop<VectorType, VectorType>(
      [&](const auto &, auto &dst, const auto &src, const auto cells) {
        for (auto cell = cells.first; cell < cells.second; cell++)
          {
            phi.reinit(cell);
            phi.gather_evaluate(src, false, true);
            for (unsigned int q = 0; q < phi.n_q_points; ++q)
              phi.submit_gradient(phi.get_gradient(q), q);
            phi.integrate_scatter(false, true, dst);
          }
      },
      dst,
      src);
  }

  const MatrixFree<dim, Number> &matrix_free;
};

template <int dim,
          int fe_degree,
          int n_points,
          int n_components,
          typename Number>
void
do_test(ConvergenceTable &table, const unsigned int n_refinements)
{
  using VectorType = Vector<Number>;

  // 1) setup triangulation
  Triangulation<dim> tria;
  GridGenerator::hyper_cube(tria);
  tria.refine_global(n_refinements);

  // 2) create dof-handler
  const FE_Q<dim>     fe_q(fe_degree);
  const FESystem<dim> fe(fe_q, n_components);

  // setup dof-handlers
  DoFHandler<dim> dof_handler(tria);
  dof_handler.distribute_dofs(fe);
  const unsigned int n_dofs = dof_handler.n_dofs();

  // 3) create constraints
  AffineConstraints<Number> constraint;
  if (false)
    {
      // hanging nodes
      DoFTools::make_hanging_node_constraints(dof_handler, constraint);

      // Dirichlet boundary
      VectorTools::interpolate_boundary_values(
        dof_handler, 0, Functions::ZeroFunction<dim>(n_components), constraint);
    }
  constraint.close();

  // 4) create matrix free
  typename MatrixFree<dim, Number>::AdditionalData additional_data;
  additional_data.mapping_update_flags = update_values | update_gradients;

  MappingQ<dim> mapping(1);
  QGauss<1>     quad(fe_degree + 1);

  MatrixFree<dim, Number> matrix_free;
  matrix_free.reinit(mapping, dof_handler, constraint, quad, additional_data);

  // 5) create linear operator
  LinearOpertor<dim, fe_degree, n_points, n_components, Number> linear_operator(
    matrix_free);

  // 6) create utility objects
  std::shared_ptr<MatrixFreeUtilities::GraphCache> cache;

  // ... vectors
  VectorType register_1(n_dofs);
  VectorType register_2(n_dofs);

  // .. pattern
  DynamicSparsityPattern dynamic_sparsity_pattern(n_dofs, n_dofs);
  DoFTools::make_sparsity_pattern(dof_handler, dynamic_sparsity_pattern);
  SparsityPattern sparsity_pattern;
  sparsity_pattern.copy_from(dynamic_sparsity_pattern);

  // full matrix
  {
    FullMatrix<Number> matrix(n_dofs, n_dofs);
    MatrixFreeUtilities::assemble_operator(
      matrix, linear_operator, sparsity_pattern, register_1, register_2, cache);
  }

  // sparse matrix
  {
    SparseMatrix<Number> matrix(sparsity_pattern);
    MatrixFreeUtilities::assemble_operator(
      matrix, linear_operator, sparsity_pattern, register_1, register_2, cache);
  }

  // 7) store results
  table.add_value("dim", dim);
  table.add_value("degree", fe_degree);
  table.add_value("n_dofs", n_dofs);
  table.add_value("n_vmults", cache->num_colors);
  table.add_value("ratio", n_dofs / cache->num_colors);

  // matrix.print_formatted (std::cout);
}

template <int dim, typename Number>
void
do_test_dim(ConvergenceTable &table)
{
  for (unsigned int i = 0; i < 4; i++)
    {
      do_test<dim, 1, 2, 1, Number>(table, i);   // linear (scalar)
      do_test<dim, 2, 3, 1, Number>(table, i);   // quadratic (scalar)
      do_test<dim, 1, 2, dim, Number>(table, i); // linear (vector)
      do_test<dim, 2, 3, dim, Number>(table, i); // quadratic (vector)
    }
}

int
main()
{
  ConvergenceTable table;

  do_test_dim<2, double>(table); // 2D
  do_test_dim<3, double>(table); // 3D

  table.write_text(std::cout);
}
```

During that I noticed following issues:
* it is not possible to leave the last optional variable away (since it is not constant)
* I would suggest that the passed in vectors I used and no copy is created.

peterrum commented:
This is the output of the program:
```
  t_full      t_sm     t_vmult   dim degree n_dofs n_vmults ratio 
5.8576e-05 1.9283e-05 1.6230e-06   2      1      4        4     1 
5.3231e-05 1.4825e-05 3.6870e-06   2      2      9        9     1 
5.0437e-05 1.2574e-05 5.2680e-06   2      1      8        8     1 
1.9993e-04 3.1407e-05 1.1136e-05   2      2     18       18     1 
3.9379e-05 1.2404e-05 3.1230e-06   2      1      9        9     1 
2.2802e-04 4.1990e-05 1.0227e-05   2      2     25       25     1 
1.4475e-04 2.8076e-05 7.8390e-06   2      1     18       18     1 
1.3121e-03 9.8274e-05 2.8424e-05   2      2     50       50     1 
1.1789e-04 2.7434e-05 6.2200e-06   2      1     25        9     2 
1.0760e-03 1.2571e-04 2.4255e-05   2      2     81       25     3 
5.2276e-04 6.0716e-05 1.8483e-05   2      1     50       18     2 
6.5667e-03 3.3469e-04 7.7812e-05   2      2    162       50     3 
4.4469e-04 7.4176e-05 1.7946e-05   2      1     81        9     9 
5.2221e-03 5.7195e-04 7.7977e-05   2      2    289       25    11 
2.4147e-03 2.0117e-04 6.2727e-05   2      1    162       18     9 
3.1118e-02 1.5982e-03 2.7735e-04   2      2    578       50    11 
5.4860e-05 1.3229e-05 3.5630e-06   3      1      8        8     1 
5.8210e-04 5.7721e-05 1.8654e-05   3      2     27       27     1 
3.9527e-04 4.4767e-05 1.8503e-05   3      1     24       24     1 
2.0820e-02 3.0421e-04 1.3103e-04   3      2     81       81     1 
2.9057e-04 6.1736e-05 1.6755e-05   3      1     27       27     1 
1.0243e-02 5.8990e-04 1.6714e-04   3      2    125      125     1 
5.3647e-03 2.3556e-04 1.0410e-04   3      1     81       81     1 
2.6221e-01 3.4065e-03 1.2597e-03   3      2    375      375     1 
2.7128e-03 2.9979e-04 9.2376e-05   3      1    125       27     4 
1.2035e-01 6.3648e-03 1.1966e-03   3      2    729      136     5 
4.4853e-02 1.2190e-03 5.0687e-04   3      1    375       81     4 
2.1050e+00 3.2598e-02 7.5359e-03   3      2   2187      408     5 
2.1515e-02 2.4958e-03 5.2142e-04   3      1    729       27    27 
1.0310e+00 1.0218e-01 7.9515e-03   3      2   4913      153    32 
3.9827e-01 1.6324e-02 4.0158e-03   3      1   2187       81    27 
2.0766e+01 7.8785e-01 6.8292e-02   3      2  14739      459    32
```
The reduction of the number of global vmults is clear. However, there is still a large gap between time to fill the sparse matrix and `n_colors`-time vmults. Could you check if there is a way to speedup the code a bit?

bangerth commented on include/deal.II/matrix_free/assemble_operator.h:
```diff
@@ -0,0 +1,278 @@
+//
+// ---------------------------------------------------------------------
+
+#pragma once
```
We tend not to use this -- it's a compile specific extension. Just use the usual include guards.

jodlbauer replied:
Changed.

bangerth commented on include/deal.II/matrix_free/assemble_operator.h:
```diff
@@ -0,0 +1,278 @@
+        typename Order = boost::typed_identity_property_map<vertices_size_type>>
+      void
+      make_coloring(const Order &&order =
+                      boost::typed_identity_property_map<vertices_size_type>{})
```
It's not illegal to have `const` rvalue references -- but they are pretty pointless :-) (I had to look this up, because I was confused. But it was a fun read: https://www.codesynthesis.com/~boris/blog/2012/07/24/const-rvalue-references/)

jodlbauer replied:
changed to const &

jodlbauer commented:
Thanks for the test case. For some reason, I had to change the parallel scheme to none for it to work properly. Otherwise I would get an error: gradients_quad_initialized.

jodlbauer commented:
Had a short look into the graph coloring, but spotted a few issues:
* only symmetric and square patterns
* only SparsityPattern supported
* does not seem to do exactly what I need (only yields 2 colors in test 2, which is certainly to few).

I could, however, move the coloring into that namespace/file, if needed.

peterrum commented:
> Thanks for the test case. For some reason, I had to change the parallel scheme to none for it to work properly. Otherwise I would get an error: gradients_quad_initialized.

Indeed, my version of deal.II is compiled without threads.

Could you also transform your block vector example into a test case? I will prepare the next days the PR to compute the matrix with the help of cell-integral kernels. I am curious what the performance differences are ;)

jodlbauer commented:
Well, the vmult approach is not intended to replace standard assembly procedures. So I strongly assume that performance of my assemble_operator will be a lot worse than any other method.

Block-Matrices need to be treated block-wise, as the BlockSparsityPattern does not allow iterating over certain rows. I could add a helper function to do that, although I am not sure how to handle the cache objects then.

peterrum commented:
> Well, the vmult approach is not intended to replace standard assembly procedures. So I strongly assume that performance of my assemble_operator will be a lot worse than any other method.
> 
> Block-Matrices need to be treated block-wise, as the BlockSparsityPattern does not allow iterating over certain rows. I could add a helper function to do that, although I am not sure how to handle the cache objects then.

I suggest that you first create a test case where you show the issues and after that we can think about improving the usability.

jodlbauer commented:
I added a test for the block-matrix case.
For simplicity, I used a Stokes Block-Matrix as operator, and "copied" it using assemble_operator.

jodlbauer commented:
> During that I noticed following issues:
>
>it is not possible to leave the last optional variable away (since it is not constant)
>I would suggest that the passed in vectors I used and no copy is created.

Do you mean the cache object? There was indeed an error in the implementation. I am still not sure how to handle the caching properly, i.e. if the cache should be returned, or only given as reference.
Also, the shared_ptr may not be necessary at all.

bangerth commented:
On 3/30/20 2:56 AM, Daniel Jodlbauer wrote:
> Had a short look into the graph coloring, but spotted a few issues:
> 
>   * only symmetric and square patterns
>   * only SparsityPattern supported
>   * does not seem to do exactly what I need (only yields 2 colors in test 2,
>     which is certainly to few).

You may be misinterpreting what the sparsity pattern there indicates. I 
believe that it suggests that if entry (i,j) exists in the sparsity pattern, 
then node i and node j of the graph collide. For your case that would mean 
that (i,j) is set in the sparsity pattern if columns i and j of the matrix you 
want to build have common entries. That also makes it clear why the sparsity 
pattern that's being used in that namespace must be symmetric -- if i and j 
collide, then j and i do as well.


> I could, however, move the coloring into that namespace/file, if needed.

Yes, ultimately that would be good.

Thanks for working on all of this!

mwichro commented:
Thanks for working on that. I've posted some time ago similar thing, but not based on sparsity pattern:
https://groups.google.com/forum/#!topic/dealii/Di_cTexTxXo
There is also direct solver  for block matrices. The internal dof renumbering works, so you can just improve assembly. Also adding initialization from block matrix would be useful, since dealii does not have direct solver for block parallel matrices.

kronbichler commented:
This PR has been idle for the past four years. Is there still interest in this direction, or are we satisfied with the improvements that were made elsewhere? I think this is a nice concept, so with some additional work this could be made a nice addition.
